sale_rental/models/sale_order.py

```diff
@@ -0,0 +1,300 @@
+        for line in self:
+            if line.rental_type == 'rental_extension':
+                if not line.extension_rental_id:
+                    raise ValidationError(_(
+                        "Missing 'Rental to Extend' on the sale order line "
+                        "with rental service %s") % line.product_id.name)
+
+                if line.rental_qty != line.extension_rental_id.rental_qty:
+                    raise ValidationError(_(
+                        "On the sale order line with rental service %s, "
+                        "you are trying to extend a rental with a rental "
+                        "quantity (%s) that is different from the quantity "
+                        "of the original rental (%s). This is not supported.")
+                        % (
+                        line.product_id.name,
+                        line.rental_qty,
+                        line.extension_rental_id.rental_qty))
+            if line.rental_type in ('new_rental', 'rental_extension'):
+                if not line.product_id.rented_product_id:
+                    raise ValidationError(_(
+                        "On the 'new rental' sale order line with product "
+                        "'%s', we should have a rental service product !") % (
+                        line.product_id.name))
+                if line.product_uom_qty !=\
+                        line.rental_qty * line.number_of_days:
+                    raise ValidationError(_(
+                        "On the sale order line with product '%s' "
+                        "the Product Quantity (%s) should be the "
+                        "number of days (%s) "
+                        "multiplied by the Rental Quantity (%s).") % (
+                        line.product_id.name, line.product_uom_qty,
+                        line.number_of_days, line.rental_qty))
+                # the module sale_start_end_dates checks that, when we have
+                # must_have_dates, we have start + end dates
+            elif line.sell_rental_id:
+                if line.product_uom_qty != line.sell_rental_id.rental_qty:
+                    raise ValidationError(_(
+                        "On the sale order line with product %s "
+                        "you are trying to sell a rented product with a "
+                        "quantity (%s) that is different from the rented "
+                        "quantity (%s). This is not supported.") % (
+                        line.product_id.name,
+                        line.product_uom_qty,
+                        line.sell_rental_id.rental_qty))
+
+    @api.multi
+    def _prepare_rental(self):
+        self.ensure_one()
+        return {'start_order_line_id': self.id}
+
+    @api.multi
+    def _get_rental_date_planned(self):
+        self.ensure_one()
+        start_date = datetime.strptime(
+            self.start_date, DEFAULT_SERVER_DATE_FORMAT)
+        return start_date.strftime(DEFAULT_SERVER_DATETIME_FORMAT)
```

Comment:
Thanks for your great work. I have just one suggestion after my review. Instead of "strptime" and "strftime" You can use functions "from_string" and "to_string" of fields.Datetime and fields.Date.